File: lib/arrow/compute/arithmetic.ex
defmodule Arrow.Compute.Arithmetic do
  alias Arrow.Array
  alias Arrow.Native

  def add(%Array{} = left, %Array{} = right), do: Native.array_compute_add(left, right)
  def divide(%Array{} = left, %Array{} = right), do: Native.array_compute_divide(left, right)
  def multiply(%Array{} = left, %Array{} = right), do: Native.array_compute_multiply(left, right)
  def subtract(%Array{} = left, %Array{} = right), do: Native.array_compute_subtract(left, right)
end

File: native/arrow_nif/src/compute/arithmetic.rs
use crate::array::{ArrayResource, XArrayRef};
use crate::impl_compute_func;
use arrow::array::{
    ArrayRef, Float32Array, Float64Array, Int16Array, Int32Array, Int64Array, Int8Array,
    UInt16Array, UInt32Array, UInt64Array, UInt8Array,
};
use arrow::datatypes::DataType;
use rustler::ResourceArc;
use std::sync::Arc;

impl_compute_func!(array_compute_add, add, arithmetic);
impl_compute_func!(array_compute_divide, divide, arithmetic);
impl_compute_func!(array_compute_multiply, multiply, arithmetic);
impl_compute_func!(array_compute_subtract, subtract, arithmetic);
